aggregator/src/app/AggregationStrategy.ts

```diff
@@ -2,216 +2,431 @@ import {
+    };
+  }
+
+  const feeTypeParts = env.FEE_TYPE.split(":");
+  assert(feeTypeParts.length === 2);
+  assert(feeTypeParts[0] === "token");
+
+  const address = feeTypeParts[1];
+  assert(/^0x[0-9a-fA-F]*$/.test(address));
+
+  assert(env.ETH_VALUE_IN_TOKENS !== nil);
+
+  return {
+    type: "token",
+    address,
+    ethValueInTokens: env.ETH_VALUE_IN_TOKENS,
+    allowLosses: env.ALLOW_LOSSES,
+    breakevenOperationCount: env.BREAKEVEN_OPERATION_COUNT,
+  };
+})();
+
+export type AggregationStrategyResult = {
+  aggregateBundle: Bundle | nil;
+  includedRows: BundleRow[];
+  bundleOverheadCost: BigNumber;
+  expectedFee: BigNumber;
+  expectedMaxCost: BigNumber;
+  failedRows: BundleRow[];
+};
+
+export type AggregationStrategyConfig =
+  typeof AggregationStrategy["defaultConfig"];
 
 export default class AggregationStrategy {
   static defaultConfig = {
     maxAggregationSize: env.MAX_AGGREGATION_SIZE,
-    fees: {
-      type: env.FEE_TYPE,
-      perGas: env.FEE_PER_GAS,
-      perByte: env.FEE_PER_BYTE,
-    },
+    fees: envFeeConfig,
   };
 
+  #tokenDecimals?: number;
+
+  // The concurrency of #checkBundle is limited by this semaphore because it can
+  // be called on many bundles in parallel
+  #checkBundleSemaphore = new Semaphore(8);
+
   constructor(
     public blsWalletSigner: BlsWalletSigner,
     public ethereumService: EthereumService,
     public config = AggregationStrategy.defaultConfig,
+    public emit: (event: AppEvent) => void = () => {},
   ) {}
 
-  async run(eligibleRows: BundleRow[]): (
-    Promise<{
-      aggregateBundle: Bundle | nil;
-      includedRows: BundleRow[];
-      failedRows: BundleRow[];
-    }>
-  ) {
+  async run(eligibleRows: BundleRow[]): Promise<AggregationStrategyResult> {
+    eligibleRows = await this.#filterRows(eligibleRows);
+
+    const bundleOverheadGas = await this.measureBundleOverheadGas();
+
     let aggregateBundle = this.blsWalletSigner.aggregate([]);
     const includedRows: BundleRow[] = [];
     const failedRows: BundleRow[] = [];
+    let expectedFee = BigNumber.from(0);
 
     while (eligibleRows.length > 0) {
       const {
         aggregateBundle: newAggregateBundle,
         includedRows: newIncludedRows,
+        expectedFees,
         failedRows: newFailedRows,
         remainingEligibleRows,
       } = await this.#augmentAggregateBundle(
         aggregateBundle,
         eligibleRows,
+        bundleOverheadGas,
       );
 
       aggregateBundle = newAggregateBundle;
       includedRows.push(...newIncludedRows);
+      expectedFee = expectedFee.add(bigSum(expectedFees));
       failedRows.push(...newFailedRows);
       eligibleRows = remainingEligibleRows;
     }
 
-    return {
-      aggregateBundle: aggregateBundle.operations.length > 0
-        ? aggregateBundle
-        : nil,
+    if (includedRows.length === 0) {
+      return {
+        aggregateBundle: nil,
+        includedRows: [],
+        bundleOverheadCost: bundleOverheadGas.mul(
```

Comment:
My brain is moving a little slow 😅 why is this the bundleOverheadGas * maxFeePerGas ?

Reply:
It's just the basic gas mechanism - transactions consume gas, and you pay a fee per unit of gas, therefore the cost is the amount of gas times the fee paid per gas.